# PredMod Daily Practice Exam Problem

Wednesday, December 31, 2025

## Problem: Regression Diagnostics and Heteroscedasticity Detection

**Course:** Predictive Modeling (Series 2-3)
**Topics:** Simple Linear Regression, Model Assumptions, Residual Diagnostics
**Difficulty:** Intermediate
**Estimated Time:** 50–70 minutes

In [2]:
import numpy as np

## **PART A: MATHEMATICAL DERIVATION \& HYPOTHESIS TESTING**

### **Scenario: Vehicle Fuel Efficiency Analysis**

A transportation logistics company conducted a study to understand the relationship between vehicle weight and fuel efficiency. Using a simple linear regression model fitted on n = 32 observations, they obtained:

$\text{MPG} = \beta_0 + \beta_1 \cdot \text{Weight} + \varepsilon$

**Model Output:**

- $\hat{\beta}_0 = 37.285$ (Intercept), $\text{se}(\hat{\beta}_0) = 1.878$
- $\hat{\beta}_1 = -5.344$ (Slope), $\text{se}(\hat{\beta}_1) = 0.559$
- Residual Standard Error (RSE) = 3.107
- R² = 0.753
- $\sum_{i=1}^{32}(x_i - \bar{x})^2 = 472.3$ (Total sum of squared deviations)
- $\bar{x} = 3.217$ (Mean weight in thousands of pounds)

In [1]:
beta0 = 37.285
se_beta0 = 1.878
beta1 = -5.344
se_beta1 = 0.559
rse = 3.107
r2 = 0.753
tss = 472.3
mean = 3.217

### **A1. Hypothesis Test and Inference**

**(a)** Perform a two-sided hypothesis test of $H_0: \beta_1 = 0$ versus $H_A: \beta_1 \neq 0$ at $\alpha = 0.05$ significance level.

- Calculate the test statistic $t^* = \frac{\hat{\beta}_1}{\text{se}(\hat{\beta}_1)}$
- Identify the rejection region using the t-distribution with appropriate degrees of freedom
- Determine the decision rule and state your conclusion in the context of the problem

In [4]:
t1 = beta1 / se_beta1
print(f"t1 = {t1:.3f}")



t1 = -9.560


**(b)** Construct a 95% confidence interval for $\beta_1$ using $\hat{\beta}_1 \pm t_{0.025, n-2} \cdot \text{se}(\hat{\beta}_1)$

- Note: Use $t_{0.025, 30} \approx 2.042$
- Interpret this interval in plain language regarding the relationship between weight and fuel consumption

**(c)** **Predictions for a specific vehicle:** Consider a vehicle weighing $x_0 = 3$ thousand pounds.

- Compute the point estimate $\hat{y}_0$ for the predicted MPG
- Calculate the 95% confidence interval for the **expected** MPG using:
$\text{se}(\hat{y}_0) = \hat{\sigma} \sqrt{\frac{1}{n} + \frac{(x_0 - \bar{x})^2}{\sum_{i=1}^{n}(x_i - \bar{x})^2}}$
- Calculate the 95% prediction interval for an **individual** MPG observation
- Explain why the prediction interval is wider than the confidence interval



***



***

### **A2. Model Assumption Violations**

The company's data analyst performed a residual analysis and discovered the following diagnostic findings:

**Tukey-Anscombe Plot observations:**

- The scatter of residuals exhibits a distinct **funnel shape**, narrowing at low fitted values and widening at high fitted values
- The smoothed LOESS curve shows systematic upward curvature across fitted values
- Residuals are not evenly distributed around the zero line in terms of vertical spread

**Scale-Location Plot observations:**

- $\sqrt{|r^*_i|}$ (square root of absolute standardized residuals) increases with fitted values
- The smoothed curve departs from the grey band of 100 simulated reference curves

***

**(a)** **Identify the violation:**

- Which regression model assumption is violated?
- State the assumption mathematically and explain what heteroscedasticity means

**(b)** **Consequences for inference:**
Explain how heteroscedasticity affects each of the following:

1. The unbiasedness of $\hat{\beta}_0$ and $\hat{\beta}_1$
2. The standard errors of the coefficient estimates
3. The validity of confidence intervals (are they still valid at their stated level?)
4. The validity of hypothesis tests and p-values
5. The interpretation of the t-statistic from part A1(a)

**(c)** **Remedial measures:**
Propose and justify **two different approaches** to address heteroscedasticity:

1. A variable **transformation** method (suggest a specific transformation and explain why it might help)
2. Describe the conceptual basis of **weighted least squares** (WLS) without detailed derivation

***



## **PART B: PYTHON IMPLEMENTATION \& MODEL DIAGNOSTICS**

### **B1. Data Simulation with Heteroscedasticity**

Write Python code to:

**(a)** Generate synthetic data that mimics the real problem with intentional heteroscedastic errors:

```python
# Pseudocode structure:
# - Generate weight values (1000 to 5000 lbs range, n=32)
# - Create errors that increase with weight: eps ~ N(0, sigma_i^2) where sigma_i depends on weight
# - Compute MPG = 37.285 - 5.344*weight + eps
# - Store in a pandas DataFrame
```

**(b)** Fit the original model and extract diagnostics:

- Regression coefficients, standard errors, and p-values
- RSE and R²
- Residuals, fitted values, standardized residuals
- Leverage values $h_i$ and Cook's distances

***

### **B2. Diagnostic Plotting**

Create a **2×2 subplot figure** containing:

1. **Top-Left: Tukey-Anscombe Plot** (Residuals vs. Fitted)
    - Scatter plot of fitted values vs. residuals
    - Add a LOESS smoothing curve in red
    - Add a horizontal reference line at y=0
    - Label axes: "Fitted Values" and "Residuals"
2. **Top-Right: Q-Q Plot** (Normal Probability Plot)
    - Plot theoretical quantiles vs. standardized residuals
    - Add a reference line y=x in red
    - Assess normality assumption
3. **Bottom-Left: Scale-Location Plot**
    - Plot fitted values vs. $\sqrt{|r^*_i|}$ (square root of absolute standardized residuals)
    - Add LOESS smoothing curve
    - Assess constant variance assumption
4. **Bottom-Right: Residuals vs. Leverage** (with Cook's distance)
    - Plot leverage statistic on x-axis, standardized residuals on y-axis
    - Add contour lines for Cook's distance at 0.5 and 1.0
    - Identify influential observations

***

### **B3. Log-Transformation Remedy**

**(a)** Fit a transformed model:
$\log(\text{MPG}) = \gamma_0 + \gamma_1 \cdot \text{Weight} + \varepsilon^*$

**(b)** Extract and compare:

- Coefficients, standard errors, p-values
- RSE and R² of the transformed model
- Create the same 2×2 diagnostic plots for the log-transformed model

**(c)** **Quantitative comparison:**

- Compare RSE values between original and log-transformed models
- Compare R² values
- Which model exhibits better constant variance in residuals?

**(d)** **Interpretation:**

- What does the coefficient $\hat{\gamma}_1$ represent in the log-transformed model?
- How would you interpret a change in weight in terms of percent change in MPG?
- Would predictions differ significantly between models?

***

## **SOLUTION SKETCH**

### **Part A Solutions:**

**A1(a) - Test Statistic:**
$t^* = \frac{-5.344 - 0}{0.559} = -9.560$

- df = 30; Critical value: $\pm 2.042$
- **Decision:** Reject $H_0$ (|-9.560| > 2.042)
- **Conclusion:** Vehicle weight has a statistically significant effect on MPG (p < 0.001)

**A1(b) - Confidence Interval:**
$-5.344 \pm 2.042 \times 0.559 = -5.344 \pm 1.141 = [-6.485, -4.203]$

- **Interpretation:** Each additional thousand pounds of weight is associated with a 4.2 to 6.5 MPG decrease on average

**A1(c) - Predictions:**

- $\hat{y}_0 = 37.285 - 5.344(3) = 21.252$ MPG
- Confidence Interval (E[Y|x₀]): Approximately [20.1, 22.4] MPG
- Prediction Interval (Y|x₀): Approximately [14.7, 27.8] MPG
- The PI is wider because it must account for both estimation uncertainty and individual observation variability

**A2(a) - Violation:** Heteroscedasticity (non-constant variance) of error terms. The assumption $\text{Var}(\varepsilon_i) = \sigma^2$ is violated; actually $\text{Var}(\varepsilon_i) = \sigma_i^2$ where $\sigma_i$ increases with fitted value.

**A2(b) - Consequences:**

1. Coefficients remain **unbiased** but inefficient
2. Standard errors are **downward-biased** (underestimated)
3. Confidence intervals are **too narrow** (nominal level < true coverage)
4. p-values are **too small**; Type I error inflation
5. The test statistic is still valid, but reported p-values are misleading

**A2(c) - Remedies:**

1. **Log-transformation:** $Y^* = \log(\text{MPG})$ often stabilizes variance, especially for rate data
2. **WLS:** Weight observations inversely by estimated variance; $\min \sum_i w_i(y_i - \hat{y}_i)^2$ where $w_i = 1/\hat{\sigma}_i^2$

***

### **Part B - Python Framework:**

```python
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.graphics.gofplots import ProbPlot
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style("whitegrid")

# 1. SIMULATE DATA WITH HETEROSCEDASTICITY
np.random.seed(42)
n = 32
weight = np.linspace(1.5, 5.4, n)  # 1500 to 5400 lbs
# Heteroscedastic errors: variance increases with weight
eps = np.random.normal(0, weight * 0.75, n)
mpg = 37.285 - 5.344 * weight + eps

df = pd.DataFrame({'Weight': weight, 'MPG': mpg})

# 2. FIT ORIGINAL MODEL
X = sm.add_constant(df['Weight'])
model_orig = sm.OLS(df['MPG'], X).fit()
print(model_orig.summary())

# Extract diagnostics
yfit = model_orig.fittedvalues
res = model_orig.resid
res_std = model_orig.resid_pearson  # Standardized residuals
infl = model_orig.get_influence()
leverage = infl.hat_matrix_diag
cooks_d = infl.cooks_distance[^0]

# 3. CREATE 2X2 DIAGNOSTIC PLOTS
fig, axes = plt.subplots(2, 2, figsize=(12, 10))

# Tukey-Anscombe (residuals vs fitted)
ax = axes[0, 0]
ax.scatter(yfit, res, alpha=0.6)
ax.axhline(y=0, color='r', linestyle='--', lw=2)
sns.regplot(x=yfit, y=res, scatter=False, lowess=True,
            line_kws={'color': 'r', 'lw': 2}, ax=ax)
ax.set_xlabel('Fitted Values')
ax.set_ylabel('Residuals')
ax.set_title('Residuals vs Fitted')

# Q-Q plot
ax = axes[0, 1]
pp = ProbPlot(res_std)
pp.qqplot(ax=ax, line='45', alpha=0.6, markersize=8)
ax.set_title('Normal Q-Q Plot')

# Scale-Location
ax = axes[1, 0]
scaled_resid = np.sqrt(np.abs(res_std))
ax.scatter(yfit, scaled_resid, alpha=0.6)
sns.regplot(x=yfit, y=scaled_resid, scatter=False, lowess=True,
            line_kws={'color': 'r', 'lw': 2}, ax=ax)
ax.set_xlabel('Fitted Values')
ax.set_ylabel('√|Standardized Residuals|')
ax.set_title('Scale-Location Plot')

# Residuals vs Leverage
ax = axes[1, 1]
ax.scatter(leverage, res_std, alpha=0.6)
ax.axhline(y=0, color='r', linestyle='--', lw=1)
# Add Cook's distance contours
from matplotlib.patches import Circle
for level in [0.5, 1.0]:
    x_coord = np.linspace(0, max(leverage), 100)
    # Approximate Cook's distance contours
ax.set_xlabel('Leverage')
ax.set_ylabel('Standardized Residuals')
ax.set_title('Residuals vs Leverage')
ax.axvline(x=2*X.shape[^1]/n, color='gray', linestyle='--', alpha=0.5)

plt.tight_layout()
plt.savefig('diagnostics_original.png', dpi=300)
plt.show()

# 4. FIT TRANSFORMED MODEL (LOG)
df['LogMPG'] = np.log(df['MPG'])
X_log = sm.add_constant(df['Weight'])
model_log = sm.OLS(df['LogMPG'], X_log).fit()
print("\n\n=== LOG-TRANSFORMED MODEL ===")
print(model_log.summary())

# Extract diagnostics for log model
yfit_log = model_log.fittedvalues
res_log = model_log.resid
res_std_log = model_log.resid_pearson

# Create same diagnostic plots for log model
# [Similar plotting code as above]

# 5. COMPARISON
print(f"\nOriginal Model - RSE: {model_orig.resid.std():.4f}, R²: {model_orig.rsquared:.4f}")
print(f"Log Model - RSE: {model_log.resid.std():.4f}, R²: {model_log.rsquared:.4f}")
```


***

## **KEY LEARNING OUTCOMES**

By completing this problem, you will:

✅ **Interpret hypothesis tests and confidence intervals** in regression context
✅ **Identify heteroscedasticity** from residual diagnostic plots
✅ **Understand consequences** of violated assumptions on inference
✅ **Apply remedial transformations** to stabilize variance
✅ **Implement comprehensive diagnostics** using Python/statsmodels
✅ **Compare models** quantitatively and qualitatively
✅ **Synthesize theory and practice** in predictive modeling

***

## **EXAM PREPARATION TIPS**

- Review Series 1 and 2 slides for hypothesis test formulas
- Practice reading residual diagnostic plots (see course exercises)
- Understand the distinction between confidence and prediction intervals
- Know when heteroscedasticity matters vs. when it's negligible
- Master the four diagnostic plots and what each reveals
- Test log and square-root transformations on your synthetic data

***

## **REFERENCE MATERIALS**

- **Series 1 (SLR):** Simple Linear Regression with Confidence/Prediction Intervals
- **Series 2 (TMA):** Testing Model Assumptions and Residual Diagnostics
- **Series 3 (MLR):** Multiple Linear Regression (extends to multivariate case)
- **Slides 1-2:** Theoretical foundations and diagnostic interpretation
- **Course Exam (Exam_PredMod.pdf):** Full solutions to similar problems

***

**Generated:** Wednesday, December 31, 2025, 08:18 CET
**Next Problem:** Thursday, January 1, 2026 (Topic: Logistic Regression \& Classification Metrics)
<span style="display:none">[^10][^11][^12][^13][^14][^15][^16][^17][^18][^19][^2][^20][^21][^22][^23][^24][^3][^4][^5][^6][^7][^8][^9]</span>

<div align="center">⁂</div>

[^1]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/589eed26-38e0-4833-8cb6-aacb6d52e2cf/12_13_series_TSSA_py.pdf

[^2]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/dcbaa214-c76d-4e28-bc5a-599a00922917/10_11_series_TSA_py.pdf

[^3]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/155a1379-6ba2-4477-becc-a42598850015/09_series_TSF_py.pdf

[^4]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/9c077f99-c827-4a37-b90a-48f0d0442235/08_series_TSMM_py.pdf

[^5]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/29672078-9b4e-47f1-a763-f241b1fd90e1/07_series_TSI_py.pdf

[^6]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/8574949b-913f-4de4-8af4-c5f688d24bc5/06_series_RT_py.pdf

[^7]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/a30faced-c139-47fe-9e90-a7dd58d53f34/05_series_LR_py.pdf

[^8]: 04_series_LMS_py.pdf

[^9]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/c18acce1-4f61-4bee-a4cd-42f9d0327537/03_series_MLR_py.pdf

[^10]: 02_series_TMA_py.pdf

[^11]: 01_series_SLR_py.pdf

[^12]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/a8e9e08f-14ce-4b12-9a6b-977c5711bd96/12_13_slides_TSSA.pdf

[^13]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/8d89bf75-0d01-44a2-b65d-a7a7ced5c187/10_11_slides_TSA.pdf

[^14]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/497c2538-5a13-4a93-a2d6-f627c6581354/09_slides_TSF.pdf

[^15]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/4d5f7484-6c43-4430-ae68-75c65d0e940b/08_slides_TSMM.pdf

[^16]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/05db51d6-2998-411f-9981-7190da646007/07_slides_TSI.pdf

[^17]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/2ad8d28e-9082-4772-ade4-f660c4144531/06_slides_RT.pdf

[^18]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/dea81977-0445-4fd9-ba56-4ac56e0f61ea/05_slides_LR.pdf

[^19]: https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/collection_47382223-78d6-4c97-9011-f0e4ab325ba5/3f94914d-692e-4f96-830e-ee6ec729acdd/04_slides_LMS.pdf

[^20]: 03_slides_MLR.pdf

[^21]: 02_slides_TMA.pdf

[^22]: main_pred_mod_py.pdf

[^23]: 01_slides_IRA_SLR.pdf

[^24]: Exam_PredMod.pdf

